firmware/thruna/DCM/Src/App/App_Regen.c

```diff
@@ -0,0 +1,80 @@
+#include "App_Regen.h"
+
+/**
+ * Check if left or right wheel is greater than 5.0km/hr
+ * @return true if wheel speed meets this condition, false
+ * otherwise
+ */
+static bool wheel_speed_in_range(void);
+
+/**
+ * Check if battery cells are less than 4.0V
+ * @return true battery cells meet this condition,
+ * false otherwise
+ */
+static bool power_limit_check(void);
+
+/**
+ * Algorithm to send negative torque request 
+ * dependent on accelerator pedal percentage
+ */
+static void compute_regen_torque_request(struct RegenBraking *regenAttr);
+
+struct RegenBraking regenAttributes;
+static float MAXREGEN = -50.0f; //TODO: find max regen torque value
+static float wheelSpeedThreshold = 5.0f;
+
+void App_Run_Regen(void) {
+    if (App_Regen_Safety(&regenAttributes)) {
+        compute_regen_torque_request(&regenAttributes);
+    } 
+    //TODO: else statement to send warning?
+
+    App_Regen_Activate(regenAttributes.left_inverter_torque, regenAttributes.right_inverter_torque);
+}
+
+bool App_Regen_Safety(struct RegenBraking *regenAttr)
+{
+    if (App_CanRx_BMS_CellTemperatures_MaxCellTemperature_Get() < 45
+    && wheel_speed_in_range() 
+    && power_limit_check()) {
+        return true;
+    }
+
+    regenAttr->left_inverter_torque = 0.0;
+    regenAttr->right_inverter_torque = 0.0;
+    return false;
```

Comment:
can we make sure this doesn't have any side effects, just handle safety in the run function, and decide whether it is safe here.

Reply:
It should be fine as two functions though and its more modular, function names would be more accurate too. I made these functions separate in case torque vectoring wants to check if they it is safe to regen without actually regenerating. Same for regen's case, idk if we'll have a case like that though. If you're referring to why i have it as two separate functions?